File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.py
class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        def dfs(node):
            isRoot = False
            isLeft = False
            isRight = False

            if not node:
                return False
            if node.val == p.val or node.val == q.val:
                isRoot = True

            isLeft = dfs(node.left)
            isRight = dfs(node.right)

            if isinstance(isLeft, TreeNode): return isLeft
            if isinstance(isRight, TreeNode): return isRight

            trues = 0
            arr = [isRoot, isLeft, isRight]
            for b in arr:
                if b: trues += 1
            if trues >= 2: return node
            return isRoot or isLeft or isRight
        
        return dfs(root)

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number[][]}
     */
    levelOrder(root) {
        let queue = []
        let res = []
        if (root !== null) {
            queue.push(root)
        }
        let level = 0
        while (queue.length > 0) {
            const levelLength = queue.length
            res[level] = []

            for (let i = 0; i < levelLength; i++) {
                const curr = queue.shift()
                res[level].push(curr.val)
                if (curr.left !== null) {
                    queue.push(curr.left)
                }
                if (curr.right !== null) {
                    queue.push(curr.right)
                }
            }
            level++
        }

        return res
    }
}
